### PR TITLE
feat(core): react authentication library with security fixes

### PR DESCRIPTION
**Pull request by eirikurn**
## What

Refactored React SPA authentication code from service-portal and application-form into a shared library. Also addresses a few reported security issues.

* Handles oidc-client and callback routes.
* Handles authentication flow with a loading screen.
* Manages user context.
* New: Renews access tokens on demand (when calling APIs) instead of continuously. This helps us support an (eg) 1-8 hour IDS session, depending on how long the user is active.
* New: Preloads a new access token sometime before it expires (when calling APIs).
* New: Monitor the IDS session and redirect to the login screen if the user is not logged in anymore.

## Why

This is security-critical code was being copied in multiple projects. We need to make sure future fixes and improvements benefit all our (non-next) react SPA projects.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/auth/react/src/lib/Authenticator/Authenticator.tsx [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        payload: user,
+      })
+    }
+
+    // This is raised when the user is signed out of the IDP.
+    const userSignedOut = () => {
+      dispatch({
+        type: ActionType.LOG_OUT,
+      })
+      if (autoLogin) {
```
Comment (atligudl) [inline libs/auth/react/src/lib/Authenticator/Authenticator.tsx:114]: Calling signIn here to redirect the user back to the IDS after logout?

**Review by atligudl — APPROVED** [review]
LGTM

File: libs/auth/react/src/lib/Authenticator/Authenticator.tsx (hunk 1 above)
Comment (eirikurn) [reply libs/auth/react/src/lib/Authenticator/Authenticator.tsx:114]: Yeah. If the user initiates a logout themself, they are sent (by the IDS) to the `post_logout_redirect_uri` (usually island.is front page).

But if the IDS session expires (1 hour) or if the user logs out in another tab, we call signin here to redirect this tab to the IDS login screen. This means:

* any sensitive information is hidden from the screen.
* the user is aware that the session is over, and doesn't try to submit changes with an expired session.
* the user can finish logging in and end up back where they were.

We'll see how this works and "feels" in practice, e.g. with delegations.

This is similar to google btw, though they take things a step further, so if I have 10 inactive tabs, all showing the login screen, it's enough for me to login to one of them, and all will then go back to where they were. That might be a cool addition here.

File: apps/application-system/form/src/main.tsx [hunk 2]
```diff
@@ -3,8 +3,10 @@ import React from 'react'
 import ReactDOM from 'react-dom'
 import * as Sentry from '@sentry/react'
 import { Integrations } from '@sentry/tracing'
-import { environment } from './environments'
+import { configure } from '@island.is/auth/react'
```
Comment (jeremybarbet) [inline apps/application-system/form/src/main.tsx:6]: Not used?

File: apps/application-system/form/src/main.tsx [hunk 3]
```diff
@@ -3,8 +3,10 @@ import React from 'react'
...
+import { ApplicationIdentityServerScope } from '@island.is/application/core'
```
Comment (jeremybarbet) [inline apps/application-system/form/src/main.tsx:7]: Not used neither?

File: libs/auth/react/src/lib/Authenticator/AuthenticatorLoadingScreen.tsx [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
-const AuthenticatorLoadingScreen: FC<{}> = () => {
+const AuthenticatorLoadingScreen = () => {
   return (
     <Box
       display="flex"
       justifyContent="center"
       alignItems="center"
       className={styles.wrapper}
+      role="progressbar"
+      aria-valuetext="Er að vinna í innskráningu"
```
Comment (jeremybarbet) [inline libs/auth/react/src/lib/Authenticator/AuthenticatorLoadingScreen.tsx:13]: Translation for that copy?

File: libs/auth/react/src/lib/Authenticator/OidcSilentSignIn.tsx [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import { getUserManager } from '../userManager'
+
+// Note: This route gets mounted as an iframe to silently log the user in
+// For further optimization it can be minimized into a barebones html file
+export const OidcSilentSignIn = (): null => {
+  useEffect(() => {
+    const userManager = getUserManager()
+    userManager.signinSilentCallback().catch(function (error) {
+      // TODO: Handle error
+      console.log(error)
```
Comment (jeremybarbet) [inline libs/auth/react/src/lib/Authenticator/OidcSilentSignIn.tsx:11]: Is it using console.log on purpose rather than the logger?

**Review by jeremybarbet — COMMENTED** [review]
Really nice, I just added a few comments

File: apps/service-portal/src/Main.tsx [hunk 6]
```diff
@@ -3,8 +3,9 @@ import React from 'react'
 import ReactDOM from 'react-dom'
 import * as Sentry from '@sentry/react'
 import { Integrations } from '@sentry/tracing'
-import { environment } from './environments'
 
+import './auth'
```
[the remark this answers is not in the document]
Comment (eirikurn) [reply apps/service-portal/src/Main.tsx:7]: Good catch. This is a mistake. Forgot to import it in application-system and remove tye old imports.

File: libs/auth/react/src/lib/Authenticator/AuthenticatorLoadingScreen.tsx (hunk 4 above)
Comment (eirikurn) [reply libs/auth/react/src/lib/Authenticator/AuthenticatorLoadingScreen.tsx:13]: That's tricky :/ 

We haven't loaded any translations at this point, while loading/authenticating.

**Review by jeremybarbet — APPROVED** [review]
LGTM

**Pull request merged**